sol/main.c

```diff
@@ -0,0 +1,507 @@
+time_resync(struct req_queue *req_queue)
+{
+	uint64_t cycles = rte_get_tsc_cycles();
+	uint64_t cycles_diff = cycles - req_queue->time_cpu_cycles;
+	uint64_t bytes_diff;
+
+	/* Compute elapsed time in bytes. */
+	bytes_diff = rte_reciprocal_divide(cycles_diff << GATEKEEPER_TIME_SHIFT,
+		req_queue->inv_cycles_per_byte);
+
+	/* Advance port time. */
+	req_queue->time_cpu_cycles = cycles;
+	req_queue->time_cpu_bytes += bytes_diff;
+	if (req_queue->time_bytes < req_queue->time_cpu_bytes)
+		req_queue->time_bytes = req_queue->time_cpu_bytes;
+}
+
+static inline void
+credits_update(struct req_queue *req_queue)
+{
+	uint64_t n_periods =
+		(req_queue->time_bytes - req_queue->tb_time_bytes) /
+		req_queue->tb_period_bytes;
+	req_queue->tb_credits += n_periods * req_queue->tb_credits_per_period;
+	req_queue->tb_credits = req_queue->tb_credits < req_queue->tb_size
+		? req_queue->tb_credits
+		: req_queue->tb_size;
+	req_queue->tb_time_bytes += n_periods * req_queue->tb_period_bytes;
+}
+
+static inline int
+credits_check(struct req_queue *req_queue, struct rte_mbuf *pkt)
+{
+	uint32_t pkt_len = pkt->pkt_len + RTE_SCHED_FRAME_OVERHEAD_DEFAULT;
+
+	if (pkt_len > req_queue->tb_credits)
+		return false;
+
+	req_queue->tb_credits -= pkt_len;
+	return true;
+}
+
+static void
+dequeue_reqs(struct sol_config *sol_conf, uint8_t tx_port,
+	uint16_t tx_queue)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	struct priority_req *entry, *next;
+
+	struct rte_mbuf *pkts_out[sol_conf->deq_burst_size];
+	uint32_t nb_pkts_out = 0;
+	uint16_t total_sent = 0;
+
+	/* Get an up-to-date view of our credits. */
+	time_resync(req_queue);
+	credits_update(req_queue);
+
+	list_for_each_entry_safe(entry, next, &req_queue->head, list) {
+		struct rte_mbuf *pkt = entry->pkt;
+
+		if (!credits_check(req_queue, pkt)) {
+			/*
+			 * XXX When under an attack, we may not want to
+			 * log this because it could become expensive.
+			 */
+			RTE_LOG(NOTICE, GATEKEEPER,
+				"sol: out of request bandwidth\n");
+			goto out;
+		}
+
+		if (req_queue->len == 1 || (entry->priority != next->priority))
+			req_queue->priorities[entry->priority] = NULL;
+		list_del(&entry->list);
+		mb_free_entry(&sol_conf->mb, entry);
+		req_queue->len--;
+
+		pkts_out[nb_pkts_out++] = pkt;
+
+		/* Advance port time. */
+		req_queue->time_bytes += pkt->pkt_len +
+			RTE_SCHED_FRAME_OVERHEAD_DEFAULT;
+
+		if (nb_pkts_out == sol_conf->deq_burst_size)
+			break;
+	}
+
+out:
+	if (list_empty(&req_queue->head)) {
+		req_queue->highest_priority = 0;
+		req_queue->lowest_priority = GK_NUM_REQ_PRIORITIES;
+	} else {
+		struct priority_req *first = list_first_entry(&req_queue->head,
+			struct priority_req, list);
+		req_queue->highest_priority = first->priority;
+	}
+
+	/* We cannot drop the packets, so re-send. */
+	do {
+		uint16_t sent = rte_eth_tx_burst(tx_port, tx_queue,
+			pkts_out + total_sent, nb_pkts_out);
+		total_sent += sent;
+		nb_pkts_out -= sent;
+	} while (nb_pkts_out > 0);
```

Comment:
It's not a bug, but this should be a `while` loop instead of a `do..while` loop because `nb_pkts_out` might be zero.